### PR TITLE
new version

### PR DESCRIPTION
**Pull request by AndroidM0nkey**

### CODE REVIEW

File: hash_map.h [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+#include <memory>
+#include <vector>
+#include <list>
+
+template<class KeyType, class ValueType, class Hash = std::hash<KeyType> >
+class HashMap {
+public:
+    using iterator = typename std::list<std::pair<const KeyType, ValueType>>::iterator;
+    using const_iterator = typename std::list<std::pair<const KeyType, ValueType>>::const_iterator;
+    using value_type = std::pair<const KeyType, ValueType>;
```
Comment (feldsherov) [inline hash_map.h]: why value_type is a pair?

File: hash_map.h [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        table.resize(sz);
+        for (iterator i = nodes.begin(); i != nodes.end(); i++) {
+            table[hasher(i->first) % sz].push_back(i);
+        }
+        size_table *= 2;
+    }
+
+    template<class ObjType>
+    void fill(ObjType obj) {
+        for (auto it: obj) {
```
Comment (feldsherov) [inline hash_map.h]: const auto&&

File: hash_map.h [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+    HashMap(Hash hasher = Hash()) : size_table(size_default), hasher(hasher), table(size_default) {}
+
+    HashMap(std::initializer_list<value_type_nonconst> init, Hash hasher = Hash()) : HashMap() {
+        fill(init);
+    }
+
+    template<class IteratorType>
+    HashMap(IteratorType bgn, IteratorType end, Hash hasher = Hash()) : HashMap() {
+        for (auto i = bgn; i != end; i++) {
```
Comment (feldsherov) [inline hash_map.h]: ```for (auto i = bgn; i != end; i++)``` calls copy constructor, that is a bad idea.

Comment (feldsherov) [reply hash_map.h]: bgn: https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

File: hash_map.h [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+            insert(*i);
+        }
+    }
+
+    std::size_t size() const {
+        return nodes.size();
+    }
+
+    bool empty() const {
+        return nodes.size() == 0;
```
Comment (feldsherov) [inline hash_map.h]: nodes.empty()

File: hash_map.h [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+
+    const ValueType &at(KeyType user_key) const {
+        if (find(user_key) == nodes.end()) {
+            throw std::out_of_range("");
+        }
+        return find(user_key)->second;
+    }
+
+    ValueType &operator[](const KeyType user_key) {
+        if (find(user_key) == nodes.end()) {
```
Comment (feldsherov) [inline hash_map.h]: You will call find three times on every call of operator[].

File: hash_map.h [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+        if (table[cur_hash].size() > size_table / size_default) {
+            rebuilt();
+        }
+    }
+
+    void erase(KeyType user_key) {
+        std::size_t cur_hash = hasher(user_key) % size_table;
+        if (find(user_key) != nodes.end()) {
+            auto cur_it = table[cur_hash].begin();
+            for (auto i : table[cur_hash]) {
```
Comment (feldsherov) [inline hash_map.h]: const auto&&

**Review by feldsherov — CHANGES_REQUESTED** [review]

File: hash_map.h (hunk 1 above)
Comment (AndroidM0nkey) [reply hash_map.h]: fixed

File: hash_map.h (hunk 2 above)
Comment (AndroidM0nkey) [reply hash_map.h]: fixed

File: hash_map.h (hunk 3 above)
Comment (AndroidM0nkey) [reply hash_map.h]: fixed with &&

File: hash_map.h (hunk 4 above)
Comment (AndroidM0nkey) [reply hash_map.h]: fixed

File: hash_map.h [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        return nodes.end();
+    }
+
+    const_iterator end() const noexcept {
+        return nodes.end();
+    }
+
+    HashMap(Hash hasher = Hash()) : size_table(size_default), hasher(hasher), table(size_default) {}
+
+    HashMap(std::initializer_list<pair_type_nonconst> init, Hash hasher = Hash()) : HashMap() {
```
Comment (feldsherov) [inline hash_map.h]: please use constructor delegation, not custom function

File: hash_map.h (hunk 5 above)
Comment (feldsherov) [reply hash_map.h]: Same comment as in the previous review.

**Review by feldsherov — CHANGES_REQUESTED** [review]

File: hash_map.h (hunk 5 above)
Comment (AndroidM0nkey) [reply hash_map.h]: fixed

File: hash_map.h (hunk 6 above)
Comment (AndroidM0nkey) [reply hash_map.h]: fixed

File: hash_map.h (hunk 7 above)
Comment (AndroidM0nkey) [reply hash_map.h]: fixed

File: hash_map.h [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+      if (table[cur_hash].size() > size_table / size_default) {
+          rebuilt();
+      }
+      return cur_it;
+
+  }
+
+  void erase(KeyType user_key) {
+      const std::size_t cur_hash = hasher(user_key) % size_table;
+      if (find(user_key) != nodes.end()) {
```
Comment (feldsherov) [inline hash_map.h:78]: Does this find necessary?

Looks like you can just iterate over all bucket, if `find(user_key) == nodes.end()`.

File: hash_map.h [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+  iterator insert(pair_type insertable) {
+      const KeyType user_key = insertable.first;
+      if (find(user_key) != nodes.end()) {
+          return find(user_key);
+      }
+      const std::size_t cur_hash = hasher(user_key) % size_table;
+      nodes.push_back(insertable);
+      const iterator cur_it = std::prev(nodes.end(), 1);
+      table[cur_hash].push_back(cur_it);
+      if (table[cur_hash].size() > size_table / size_default) {
```
Comment (feldsherov) [inline hash_map.h:69]: Condition is strange. Usually, you want to rehash if the number of elements is twice bigger than the number of buckets or smth like that.

**Review by feldsherov — APPROVED** [review]
Code OK. But read my comments, please.

**Pull request merged**